lambdas/s3hash/src/t4_lambda_s3hash/__init__.py

```diff
@@ -0,0 +1,388 @@
+        return {
+            "Bucket": self.bucket,
+            "Key": self.key,
+        }
+
+
+class MPURef(pydantic.BaseModel):
+    bucket: str
+    key: str
+    id: str
+
+    @property
+    def boto_args(self):
+        return {
+            "Bucket": self.bucket,
+            "Key": self.key,
+            "UploadId": self.id,
+        }
+
+
+# 8 MiB -- boto3 default: https://boto3.amazonaws.com/v1/documentation/api/latest/reference/customizations/s3.html#boto3.s3.transfer.TransferConfig
+MIN_PART_SIZE = 1024**2 * 8
+MAX_PARTS = 10000  # Maximum number of parts per upload supported by S3
+
+
+# XXX: import this logic from quilt3 when it's available
+def get_part_size(file_size: int) -> T.Optional[int]:
+    if file_size < MIN_PART_SIZE:
+        return None  # use single-part upload (and plain SHA256 hash)
+
+    # NOTE: in the case where file_size is exactly equal to MIN_PART_SIZE,
+    # boto creates a 1-part multipart upload :shrug:
+    part_size = MIN_PART_SIZE
+    num_parts = math.ceil(file_size / part_size)
+
+    while num_parts > MAX_PARTS:
+        part_size *= 2
+        num_parts = math.ceil(file_size / part_size)
+
+    return part_size
+
+
+async def get_existing_checksum(
+    location: S3ObjectSource,
+    parts: T.Sequence[PartDef],
+) -> T.Optional[Checksum]:
+    # XXX
+    # get object attributes and see if part sizes and checksum algo matches our standard
+    # return the existing checksum if it's compliant
+    # otherwise return None
+    return None
+
+
+def hash_parts(parts: T.Sequence[bytes]) -> bytes:
+    return hashlib.sha256(b"".join(parts)).digest()
+
+
+class PartDef(pydantic.BaseModel):
+    part_number: int
+    range: T.Optional[T.Tuple[int, int]]
+
+    @property
+    def boto_args(self):
+        args: T.Dict[str, T.Any] = {"PartNumber": self.part_number}
+        if self.range:
+            args["CopySourceRange"] = f"bytes={self.range[0]}-{self.range[1]}"
+        return args
+
+    @classmethod
+    def for_range(cls, part_number: int, offset: int, size: int):
+        return cls(
+            part_number=part_number,
+            range=(offset, offset + size - 1),
+        )
+
+
+PARTS_SINGLE = [PartDef(part_number=1, range=None)]
+
+
+async def get_parts_for_location(location: S3ObjectSource) -> T.List[PartDef]:
+    total_size = (await S3.get().head_object(**location.boto_args))["ContentLength"]
+    part_size = get_part_size(total_size)
+
+    # single-part upload
+    if part_size is None:
+        return PARTS_SINGLE
+
+    # multipart upload
+    offset = 0
+    part_number = 1
+    parts = []
+    while offset < total_size:
+        actual_part_size = min(part_size, total_size - offset)
+        parts.append(
+            PartDef.for_range(
+                part_number=part_number,
+                offset=offset,
+                size=actual_part_size,
+            )
+        )
+        offset += actual_part_size
+        part_number += 1
+
+    return parts
+
+
+@tenacity.retry
```

Comment:
Maybe configure [boto's retries](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/retries.html) instead?

Reply:
yeah, looks promising, i didn't know about it